File: sdk/typescript/src/clients/repos.ts
/**
 * Repositories resource client.
 *
 * Design Reference: DR-5
 * Requirements: 7.1, 7.2, 7.3, 7.4
 */

import type { HTTPTransport } from '../transport.js';
import type { Repository } from '../types/index.js';

/**
 * Parse ISO date string to Date object.
 */
function parseDate(dateStr: string): Date {
  return new Date(dateStr.replace(/Z$/, '+00:00'));
}

/**
 * Get value from object, trying camelCase first then snake_case.
 */
function getValue<T>(
  data: Record<string, unknown>,
  camel: string,
  snake: string,
  defaultValue?: T
): T {
  if (camel in data) {
    return data[camel] as T;
  }
  return (data[snake] ?? defaultValue) as T;
}

/**
 * Parse repository data handling both camelCase and snake_case.
 */
function parseRepository(data: Record<string, unknown>): Repository {
  const createdAtVal = getValue<string>(data, 'createdAt', 'created_at', '');
  return {
    repoId: getValue<string>(data, 'repoId', 'repo_id', ''),
    name: data.name as string,
    ownerId: getValue<string>(data, 'ownerId', 'owner_id', ''),
    ownerName: getValue<string | null>(data, 'ownerName', 'owner_name', null),
    description: (data.description as string | null) ?? null,
    visibility: data.visibility as 'public' | 'private',
    defaultBranch: getValue<string>(data, 'defaultBranch', 'default_branch', 'main'),
    cloneUrl: getValue<string>(data, 'cloneUrl', 'clone_url', ''),
    starCount: getValue<number>(data, 'starCount', 'stars', 0),
    createdAt: parseDate(createdAtVal),
  };
}

/**
 * Client for repository-related operations.
 *
 * Design Reference: DR-5
 */
export class ReposClient {
  private transport: HTTPTransport;

  /**
   * Initialize the repos client.
   *
   * @param transport - HTTP transport for making requests
   */
  constructor(transport: HTTPTransport) {
    this.transport = transport;
  }

  /**
   * Create a new repository.
   *
   * @param name - Repository name
   * @param description - Optional repository description
   * @param visibility - "public" or "private" (default: "public")
   * @returns Repository object with repoId, cloneUrl, etc.
   * @throws AuthenticationError if signature is invalid
   * @throws ConflictError if repository already exists
   *
   * Requirements: 7.1, 7.2
   */
  async create(
    name: string,
    description?: string | null,
    visibility: 'public' | 'private' = 'public'
  ): Promise<Repository> {
    const body: Record<string, unknown> = {
      name,
      description: description ?? null,
      visibility,
    };

    const response = await this.transport.signedRequest(
      'POST',
      '/v1/repos',
      'repo_create',
      body
    );

    const data = (response.data ?? {}) as Record<string, unknown>;
    return parseRepository(data);
  }

  /**
   * Get repository information.
   *
   * @param repoId - The unique repository identifier
   * @returns Repository object with metadata including starCount
   * @throws NotFoundError if repository not found
   *
   * Requirements: 7.3
   */
  async get(repoId: string): Promise<Repository> {
    const response = await this.transport.unsignedRequest(
      'GET',
      `/v1/repos/${repoId}`
    );

    const data = (response.data ?? {}) as Record<string, unknown>;
    return parseRepository(data);
  }

  /**
   * List repositories owned by the authenticated agent.
   *
   * @returns List of Repository objects
   *
   * Requirements: 7.4
   */
  async list(): Promise<Repository[]> {
    const response = await this.transport.signedRequest(
      'GET',
      '/v1/repos',
      'repo_list',
      {}
    );

    const data = (response.data ?? {}) as Record<string, unknown>;
    const repos = (data.repos ?? []) as Record<string, unknown>[];
    return repos.map(parseRepository);
  }
}

File: sdk/rust/src/clients/repos.rs
//! Repositories resource client.
//!
//! Design Reference: DR-5
//! Requirements: 7.1, 7.2, 7.3, 7.4

use std::collections::HashMap;
use std::sync::Arc;

use serde_json::Value;

use crate::error::Error;
use crate::transport::HttpTransport;
use crate::types::Repository;

/// Client for repository-related operations.
pub struct ReposClient {
    transport: Arc<HttpTransport>,
}

impl ReposClient {
    /// Create a new repos client.
    pub fn new(transport: Arc<HttpTransport>) -> Self {
        Self { transport }
    }

    /// Create a new repository.
    ///
    /// # Arguments
    ///
    /// * `name` - Repository name
    /// * `description` - Optional repository description
    /// * `visibility` - "public" or "private" (default: "public")
    ///
    /// # Returns
    ///
    /// Repository object with repo_id, clone_url, etc.
    ///
    /// # Errors
    ///
    /// Returns an error if creation fails.
    ///
    /// Requirements: 7.1, 7.2
    pub async fn create(
        &self,
        name: &str,
        description: Option<&str>,
        visibility: Option<&str>,
    ) -> Result<Repository, Error> {
        let mut body: HashMap<String, Value> = HashMap::new();
        body.insert("name".to_string(), Value::String(name.to_string()));
        body.insert(
            "description".to_string(),
            description.map_or(Value::Null, |d| Value::String(d.to_string())),
        );
        body.insert(
            "visibility".to_string(),
            Value::String(visibility.unwrap_or("public").to_string()),
        );

        let response: Value = self
            .transport
            .signed_request("POST", "/v1/repos", "repo_create", body)
            .await?;

        let data = response
            .get("data")
            .ok_or_else(|| Error::Http("Missing data in response".to_string()))?;

        serde_json::from_value(data.clone()).map_err(Error::from)
    }

    /// Get repository information.
    ///
    /// # Arguments
    ///
    /// * `repo_id` - The unique repository identifier
    ///
    /// # Returns
    ///
    /// Repository object with metadata including star_count
    ///
    /// # Errors
    ///
    /// Returns an error if the repository is not found.
    ///
    /// Requirements: 7.3
    pub async fn get(&self, repo_id: &str) -> Result<Repository, Error> {
        let response: Value = self
            .transport
            .unsigned_request::<Value>("GET", &format!("/v1/repos/{repo_id}"), None, None::<&()>)
            .await?;

        let data = response
            .get("data")
            .ok_or_else(|| Error::Http("Missing data in response".to_string()))?;

        serde_json::from_value(data.clone()).map_err(Error::from)
    }

    /// List repositories owned by the authenticated agent.
    ///
    /// # Returns
    ///
    /// List of Repository objects
    ///
    /// # Errors
    ///
    /// Returns an error if the request fails.
    ///
    /// Requirements: 7.4
    pub async fn list(&self) -> Result<Vec<Repository>, Error> {
        let response: Value = self
            .transport
            .signed_request("GET", "/v1/repos", "repo_list", HashMap::new())
            .await?;

        let data = response
            .get("data")
            .ok_or_else(|| Error::Http("Missing data in response".to_string()))?;

        let repos = data
            .get("repos")
            .ok_or_else(|| Error::Http("Missing repos in response".to_string()))?;

        serde_json::from_value(repos.clone()).map_err(Error::from)
    }
}
